### PR TITLE
Solana provider request

### PR DESCRIPTION
**Pull request by darkdh**
Resolves https://github.com/brave/brave-browser/issues/22974
1. Integrate requests for all methods with API implementation
2. Handle result difference between API and request like `signTransaction` and `signAllTransactions`
3. Error and error message should be propagated
## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [x] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [x] New files have MPL-2.0 license header
- [x] Adequate test coverage exists to prevent regressions
- [x] Major classes, functions and non-trivial code blocks are well-commented
- [x] Changes in component dependencies are properly reflected in `gn`
- [x] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [x] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_wallet/browser/solana_provider_impl.cc [hunk 1]
```diff
@@ -496,14 +590,59 @@ void SolanaProviderImpl::OnSignMessageRequestProcessed(
...
+    mojom::SolanaProviderError error,
+    const std::string& error_message,
+    const std::vector<std::vector<uint8_t>>& serialized_txs) {
+  base::Value result(base::Value::Type::DICTIONARY);
+  if (error == mojom::SolanaProviderError::kSuccess) {
+    base::Value signatures(base::Value::Type::LIST);
+    for (const auto& serialized_tx : serialized_txs) {
+      auto tx = SolanaTransaction::FromSignedTransactionBytes(serialized_tx);
+      DCHECK(tx);
+      if (!result.GetDict().contains(kPublicKey))
```
Comment (yrliou) [inline components/brave_wallet/browser/solana_provider_impl.cc:637]: Please help add a comment here to say something like
```
// The API expect only one signer (the selected account) from all transactions in the array.
```

File: components/brave_wallet/browser/solana_transaction.h [hunk 2]
```diff
@@ -69,6 +75,7 @@ class SolanaTransaction {
  private:
   FRIEND_TEST_ALL_PREFIXES(SolanaTransactionUnitTest, GetBase64EncodedMessage);
   SolanaMessage message_;
+  std::vector<uint8_t> signatures_;
```
Comment (yrliou) [inline components/brave_wallet/browser/solana_transaction.h:78]: Please put a comment to say this is only assigned when `FromSignedTransactionBytes` is called.

**Review by yrliou — APPROVED** [review]
LGTM, just a few nits.

File: components/brave_wallet/browser/solana_provider_impl.cc (hunk 1 above)
Comment (darkdh) [reply components/brave_wallet/browser/solana_provider_impl.cc]: fixed and squashed in 88c253f4b44380b67a6c8c3201f51e04d31e9d4e

File: components/brave_wallet/browser/solana_transaction.h (hunk 2 above)
Comment (darkdh) [reply components/brave_wallet/browser/solana_transaction.h:79]: fixed and squashed in 88c253f4b44380b67a6c8c3201f51e04d31e9d4e

File: components/brave_wallet/browser/solana_provider_impl.cc [hunk 3]
```diff
@@ -420,10 +424,100 @@ void SolanaProviderImpl::SignMessage(
...
+    std::move(callback).Run(mojom::SolanaProviderError::kParsingError,
+                            l10n_util::GetStringUTF8(IDS_WALLET_PARSING_ERROR),
+                            base::Value(base::Value::Type::DICTIONARY));
+    return;
+  }
+
+  if (*method == solana::kConnect) {
+    absl::optional<base::Value> option = absl::nullopt;
+    if (params)
+      option = base::Value(params->Clone());
```
[the remark this answers is not in the document]
Comment (darkdh) [reply components/brave_wallet/browser/solana_provider_impl.cc]: fixed and squashed in 45c0d0cab70deb53eb11a7ef7604429cf23cfddd

**Pull request merged**